operator/api/v1alpha2/seldondeployment_webhook.go

```diff
@@ -0,0 +1,446 @@
+			}
+
+		}
+	}
+}
+
+// -----
+
+func addDefaultsToGraph(pu *PredictiveUnit) {
+	if pu.Type == nil {
+		ty := UNKNOWN_TYPE
+		pu.Type = &ty
+	}
+	if pu.Implementation == nil {
+		im := UNKNOWN_IMPLEMENTATION
+		pu.Implementation = &im
+	}
+	for i := 0; i < len(pu.Children); i++ {
+		addDefaultsToGraph(&pu.Children[i])
+	}
+}
+
+func (r *SeldonDeployment) DefaultSeldonDeployment() {
+
+	var firstPuPortNum int32 = 9000
+	if env_preditive_unit_service_port, ok := os.LookupEnv("PREDICTIVE_UNIT_SERVICE_PORT"); ok {
+		portNum, err := strconv.Atoi(env_preditive_unit_service_port)
+		if err != nil {
+			seldondeploymentlog.Error(err, "Failed to decode PREDICTIVE_UNIT_SERVICE_PORT will use default 9000", "value", env_preditive_unit_service_port)
+		} else {
+			firstPuPortNum = int32(portNum)
+		}
+	}
+	nextPortNum := firstPuPortNum
+
+	portMap := map[string]int32{}
+
+	if r.ObjectMeta.Namespace == "" {
+		r.ObjectMeta.Namespace = "default"
+	}
+
+	for i := 0; i < len(r.Spec.Predictors); i++ {
+		p := r.Spec.Predictors[i]
+		if p.Graph.Type == nil {
+			ty := UNKNOWN_TYPE
+			p.Graph.Type = &ty
+		}
+		// Add version label for predictor if not present
+		if p.Labels == nil {
+			p.Labels = map[string]string{}
+		}
+		if _, present := p.Labels["version"]; !present {
+			p.Labels["version"] = p.Name
+		}
+		addDefaultsToGraph(p.Graph)
+
+		fmt.Println("predictor is now")
+		jstr, _ := json.Marshal(p)
+		fmt.Println(string(jstr))
+
+		r.Spec.Predictors[i] = p
+
+		for j := 0; j < len(p.ComponentSpecs); j++ {
+			cSpec := r.Spec.Predictors[i].ComponentSpecs[j]
+
+			// add service details for each container - looping this way as if containers in same pod and its the engine pod both need to be localhost
+			for k := 0; k < len(cSpec.Spec.Containers); k++ {
+				con := &cSpec.Spec.Containers[k]
+
+				if _, present := portMap[con.Name]; !present {
+					portMap[con.Name] = nextPortNum
+					nextPortNum++
+				}
+				portNum := portMap[con.Name]
+
+				pu := GetPredcitiveUnit(p.Graph, con.Name)
+
+				if pu != nil {
+
+					if pu.Endpoint == nil {
+						pu.Endpoint = &Endpoint{Type: REST}
+					}
+					var portType string
+					if pu.Endpoint.Type == GRPC {
+						portType = "grpc"
+					} else {
+						portType = "http"
+					}
+
+					if con != nil {
+						existingPort := GetPort(portType, con.Ports)
+						if existingPort != nil {
+							portNum = existingPort.ContainerPort
+						}
+					}
+
+					// Set ports and hostname in predictive unit so engine can read it from SDep
+					// if this is the first componentSpec then it's the one to put the engine in - note using outer loop counter here
+					if _, hasSeparateEnginePod := r.Spec.Annotations[ANNOTATION_SEPARATE_ENGINE]; j == 0 && !hasSeparateEnginePod {
+						pu.Endpoint.ServiceHost = "localhost"
+					} else {
+						containerServiceValue := GetContainerServiceName(r, p, con)
+						pu.Endpoint.ServiceHost = containerServiceValue + "." + r.ObjectMeta.Namespace + ".svc.cluster.local."
```

Comment:
".svc.cluster.local." is making assumption that the cluster domain suffix is cluster.local, but it is not always true.

Reply:
We now have #891 to fix this later